### PR TITLE
Escape '/' characters when creating k8s names for classes and plans

### PR DESCRIPTION
**Pull request by jberkhahn**
Kubernetes names cannot contain / characters. This PR replaces those with the utf-8 hex code for /.

### CODE REVIEW

File: pkg/controller/controller_test.go [hunk 1]
```diff
@@ -1852,6 +1855,88 @@ func TestCatalogConversionClusterServicePlanBindable(t *testing.T) {
...
+    {
+      "name": "mysql",
+      "id": "mysql/1234",
+      "plans": [{
+        "name": "small-mysql",
+        "id": "mysql/100mb"
+      },
+      {
+        "name": "big-mysql",
+        "id": "mysql/100gb"
```
Comment (duglin) [inline pkg/controller/controller_test.go]: add tests that will show conflicts if the escaping isn't done right. For example: mysql/foo & mysql2ffoo for the current PR will show a problem.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -866,11 +867,25 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
 
+func generateEscapedName(externalID string) string {
+	re := regexp.MustCompile("[^a-zA-Z0-9]")
+
+	escapedName := re.ReplaceAllStringFunc(externalID, func(in string) string {
+		return fmt.Sprintf("-%x-", in)
+	})
+
+	if len(escapedName) > 63 {
+		escapedName = escapedName[0:62]
```
Comment (duglin) [inline pkg/controller/controller.go]: not thrilled but not sure what else to do - aside randomizing part of the string, but we can hold off on something that drastic until later I guess.

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -866,11 +867,25 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
 	return serviceClasses, servicePlans, nil
 }
 
+func generateEscapedName(externalID string) string {
+	re := regexp.MustCompile("[^a-zA-Z0-9]")
+
+	escapedName := re.ReplaceAllStringFunc(externalID, func(in string) string {
+		return fmt.Sprintf("-%x-", in)
```
Comment (duglin) [inline pkg/controller/controller.go]: why do you have the trailing "-"?  If you use "-%0x" then you should always get a 2 digit number and don't need an end marker.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -912,12 +927,16 @@ func convertAndFilterCatalog(in *osb.CatalogResponse, restrictions *v1beta1.Cata
 			}
 			serviceClass.Spec.ExternalMetadata = &runtime.RawExtension{Raw: metadata}
 		}
-		serviceClass.SetName(svc.ID)
+		if existingServiceClasses[svc.ID] != nil {
+			serviceClass.SetName(existingServiceClasses[svc.ID].Name)
+		} else {
+			serviceClass.SetName(generateEscapedName(svc.ID))
+		}
```
[the remark this answers is not in the document]
Comment (jberkhahn) [reply pkg/controller/controller.go]: yeah, we can't change names of pre existing resources in old deployments that are upgraded

File: pkg/controller/controller.go (hunk 3 above)
Comment (jberkhahn) [reply pkg/controller/controller.go]: I think this is a bit more complicated than that. The first number you provide is the default width, but as the value that is actually coming out is really a hex number, you need to provide a precision value to truncate it to two characters. I think you actualy want "-%0.1x", but I am not sure about that.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -866,11 +867,25 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
 }
 
+func generateEscapedName(externalID string) string {
+	re := regexp.MustCompile("[^a-zA-Z0-9]")
+
+	escapedName := re.ReplaceAllStringFunc(externalID, func(in string) string {
+		return fmt.Sprintf("-%x-", in)
+	})
+
+	if len(escapedName) > 63 {
```
[the remark this answers is not in the document]
Comment (jberkhahn) [reply pkg/controller/controller.go]: none of the const in the main kube validation package are public, so should i just declare our own somewhere and use that?

[the remark this answers is not in the document]
Comment (duglin) [reply pkg/controller/controller.go]: re: hashing - yep. Jonathan and I were chatting about this the other day and I liked the idea of chopping the string when it's too long and replacing the tail of end it with a hash. At least then for cases where it's not too long they still get something readable.

File: pkg/controller/controller.go (hunk 2 above)
Comment (duglin) [reply pkg/controller/controller.go]: re: hashing - yep. Jonathan and I were chatting about this the other day and I liked the idea of chopping the string when it's too long and replacing the tail of end it with a hash. At least then for cases where it's not too long they still get something readable.

File: pkg/controller/controller.go (hunk 3 above)
Comment (jberkhahn) [reply pkg/controller/controller.go]: Those don't work for actually weird characters like •

Comment (MHBauer) [reply pkg/controller/controller.go]: ya, ID is globally-unique-string, which, assuming an character coding greater than ascii, means there are a lot of variable width stuff that can come out.

I don't get your output.

https://play.golang.org/p/v3mX7jtpgMG
returns
```
61  61  61
2009  09  09
```

for me, and if we expand it 
https://play.golang.org/p/Q76JiQw_HzH
```
package main

import "fmt"

func main() {
	printer("a")
	printer("\t")
	printer("t")
	printer("益")
	printer("🙈")
}

func printer(s string) {
	fmt.Printf("%0x  %2x  %0.1x\n", s, s, s) // tabs
}
```
yields
```
61  61  61
09  09  09
e79b8a  e79b8a  e7
f09f9988  f09f9988  f0
```
(I think tab maybe gets converted differently depending on where it's parsed?)

File: pkg/controller/controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply pkg/controller/controller.go]: Some of these look public. 
```
$ rg '= 63' vendor/k8s.io/apimachinery/pkg/util/validation

vendor/k8s.io/apimachinery/pkg/util/validation/validation.go
34:const qualifiedNameMaxLength int = 63
90:const LabelValueMaxLength int = 63
110:const DNS1123LabelMaxLength int = 63
148:const DNS1035LabelMaxLength int = 63
```
We should probably have a more project-local const (even if it's a reassignment of a public const elsewhere), but if you ever want it to fail if it's changed, maybe our own hardcoded bit that checks those other bits.

[the remark this answers is not in the document]
Comment (MHBauer) [reply pkg/controller/controller.go]: I'm not sure the 63 limit is relevant for us, but I haven't tried anything long in a while.

File: pkg/controller/controller_clusterservicebroker.go [hunk 6]
```diff
@@ -246,6 +232,20 @@ func (c *controller) reconcileClusterServiceBroker(broker *v1beta1.ClusterServic
...
+			}
+			return err
+		}
+
+		// convert the broker's catalog payload into our API objects
+		glog.V(4).Info(pcb.Message("Converting catalog response into service-catalog API"))
+
 		// reconcile the serviceClasses that were part of the broker's catalog
 		// payload
 		for _, payloadServiceClass := range payloadServiceClasses {
```
[the remark this answers is not in the document]
Comment (MHBauer) [reply pkg/controller/controller_clusterservicebroker.go]: new changes below make sense to me.

**Comment by MHBauer** [comment]
Overall, maybe we can dump the 63 limit, but it needs to be tested.

File: pkg/controller/controller.go (hunk 3 above)
Comment (duglin) [reply pkg/controller/controller.go]: I had an extra space before the first tab by mistake when I copied it into the comment - that's why you see the "20".

I think the `%2x` output seems fine since this is meant to be something people type of the cmd line and I sure hope they would avoid really odd chars like that when the rest of the line should be ascii.  However, I'm also fine with being really really worried and using @jberkhahn's original proposal of starting and ending with "-" to avoid possible conflicts.  Let's just pick one and go with it!

**Comment by jberkhahn** [comment]
@MHBauer @duglin new revision up, tell me what you think. I've tried to cover all the corner cases I can think of in the new tests. Let me know if you can think of anything that screws up or causes a dupe. I'm also only using md5 as the hashing algorithm because i wanted something short enough to leave a meaningful amount of the original ID so hopefully its something readable.

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -866,11 +868,46 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
 	return serviceClasses, servicePlans, nil
 }
 
+func generateEscapedName(externalID string) string {
+	startingRE := regexp.MustCompile("^\\.")
+	re := regexp.MustCompile("[^a-z0-9.]")
+	endingRE := regexp.MustCompile("\\.$")
+	invalidHyphens := regexp.MustCompile("--")
+	invalidHyphenPeriod := regexp.MustCompile("-\\.")
+	invalidPeriodHyphen := regexp.MustCompile("\\.-")
```
Comment (MHBauer) [inline pkg/controller/controller.go]: a paragraph explaining all of this in words would be nice. or a bulleted list of the cases that each of these cover.

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -866,11 +870,56 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
 	return serviceClasses, servicePlans, nil
 }
 
+// generateEscapedName takes in an OSB ID and filters
+// it to fit K8S name restrictions. It escapes all
+// characters except for lowercase alphanumerics and
+// periods, chaning them into their UTF-8 hex
```
Comment (duglin) [inline pkg/controller/controller.go]: s/chaning/changing/

**Comment by MHBauer** [comment]
That was strange, I wonder what happened?

Remind me to review this again tomorrow.

**Comment by jberkhahn** [comment]
Ok, I think this one is finally ready for review.

@MHBauer want to take another look at it?

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -863,11 +873,64 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+// of the escaped name, with trailing periods trimmed.
+func generateEscapedName(externalID string) string {
+	buffer := bytes.Buffer{}
+	lenOrigin := len(externalID)
+	prevDot := false
+	prevDash := false
+
+	for i, ch := range externalID {
+		// Don't need to escape [a-z0-9.]
+		// '.' isn't ok at start, end or after '-', so escape those cases
```
Comment (MHBauer) [inline pkg/controller/controller.go]: nice comment :+1:

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -863,11 +873,64 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+		if (ch >= 'a' && ch <= 'y') || (ch >= '0' && ch <= '9') {
+
```
Comment (MHBauer) [inline pkg/controller/controller.go]: extra newline, I think?

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -863,11 +873,64 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+			buffer.WriteString(string(ch))
+			prevDash = false
+			prevDot = true
+		} else if ch == '-' && i != 0 && i != lenOrigin-1 && !prevDot {
+			buffer.WriteString(string(ch))
+			prevDash = true
+			prevDot = false
+		} else {
+			start, end := "z", "z" // By default start/end escaped ch with 'z'
+
```
Comment (MHBauer) [inline pkg/controller/controller.go]: extra line?

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -863,11 +873,64 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+			buffer.WriteString(fmt.Sprintf("%s%x%s", start, ch, end)) // append
```
Comment (MHBauer) [inline pkg/controller/controller.go]: this isn't different from any other writestring calls, so I don't think we need the append comment.

File: pkg/controller/controller.go [hunk 13]
```diff
@@ -863,11 +873,64 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+// generateEscapedName takes in an OSB ID and filters
+// it to fit K8S name restrictions. It escapes all
+// characters except for lowercase alphanumerics (excluding z
+// as we use that for the escape character), hyphens, and
+// periods, changing them into their UTF-8 hex
+// equivalents separated by 'z'. Hyphens that come after periods and
+// periods that come after hyphens or periods are also escaped.
+// If the resulting string is over the K8S limit of 63 characters,
+// it appends the md5 hash of the original OSB ID to the first 31 characters
+// of the escaped name, with trailing periods trimmed.
```
Comment (MHBauer) [inline pkg/controller/controller.go]: not clear until reading later code that the trailing periods are that on the 31 chars, not the chars + md5.

File: pkg/controller/controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MHBauer) [reply pkg/controller/controller.go]: needs a comment that that's what the if is for.

File: pkg/controller/controller_serviceclass_test.go [hunk 14]
```diff
@@ -24,10 +24,11 @@ import (
 	"github.com/kubernetes-incubator/service-catalog/test/fake"
 	"k8s.io/apimachinery/pkg/runtime"
 
+	"reflect"
```
Comment (MHBauer) [inline pkg/controller/controller_serviceclass_test.go]: if we're moving things around, this should go up to the top.

File: pkg/controller/controller_serviceclass_test.go [hunk 15]
```diff
@@ -24,10 +24,11 @@ import (
...
+
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/labels"
 	clientgotesting "k8s.io/client-go/testing"
```
Comment (MHBauer) [inline pkg/controller/controller_serviceclass_test.go]: and the k8s.io stuff should be merged.

File: pkg/controller/controller_serviceclass_test.go [hunk 16]
```diff
@@ -24,10 +24,11 @@ import (
 	"github.com/kubernetes-incubator/service-catalog/test/fake"
```
Comment (MHBauer) [inline pkg/controller/controller_serviceclass_test.go]: and all the internal packages should be at the bottom.

**Comment by duglin** [comment]
The escaping algorithm looks good now - thanks!

File: pkg/controller/controller_test.go [hunk 17]
```diff
@@ -1852,6 +1907,232 @@ func TestCatalogConversionClusterServicePlanBindable(t *testing.T) {
...
+				"id": "thisnameisreallyreallywaytool-onghowcouldanyonepossiblyreadthisplansname"
+      },
+			{
+				"name": "too-long-with-problem-period",
+				"id": "thisnameisreallyreallywaytool.onghowcouldanyonepossiblyreadthisplansname"
+			}]
+    }
+]}`
+
+func TestCatalogConversionInvalidCharactersInExternalID(t *testing.T) {
```
Comment (jboyd01) [inline pkg/controller/controller_test.go]: This is the meat of testing generateEscapedName() right?  For me it borders on incomprehensible for verifying escaped strings are as expected.  How about adding unit tests specific to only generateEscapedName()?

**Comment by jboyd01** [comment]
Thanks for the new tests @jberkhahn 
/lgtm

File: pkg/controller/controller_test.go [hunk 18]
```diff
@@ -61,15 +61,15 @@ import (
...
 
 const (
-	testClusterServiceClassGUID            = "CSCGUID"
-	testClusterServicePlanGUID             = "CSPGUID"
-	testNonbindableClusterServiceClassGUID = "UNBINDABLE-CLUSTERSERVICECLASS"
-	testNonbindableClusterServicePlanGUID  = "UNBINDABLE-CLUSTERSERVICEPLAN"
-	testServiceInstanceGUID                = "IGUID"
-	testServiceBindingGUID                 = "BGUID"
+	testClusterServiceClassGUID            = "cscguid"
+	testClusterServicePlanGUID             = "cspguid"
```
Comment (MHBauer) [inline pkg/controller/controller_test.go]: lot of places that these constants can be used instead of changing the hardcoded strings.

File: pkg/controller/controller.go [hunk 19]
```diff
@@ -863,11 +874,63 @@ func convertAndFilterCatalogToNamespacedTypes(namespace string, in *osb.CatalogR
...
+// it to fit K8S name restrictions. It escapes all
+// characters except for lowercase alphanumerics (excluding z
+// as we use that for the escape character), hyphens, and
+// periods, changing them into their UTF-8 hex
+// equivalents separated by 'z'. Hyphens that come after periods and
+// periods that come after hyphens or periods are also escaped.
+// If the resulting string is over the K8S limit of 63 characters,
+// it appends the md5 hash of the original OSB ID to the first 31 characters
+// (minus trailing periods) of the escaped name.
+func GenerateEscapedName(externalID string) string {
```
Comment (MHBauer) [inline pkg/controller/controller.go]: Do we want to export this? None of the test packages look like they are the separate _test package type. I suppose someone might want to import this to pre-prep their service names?

Comment (jberkhahn) [reply pkg/controller/controller.go]: @jboyd01 wanted this to be unit tested, so I made it public

Comment (MHBauer) [reply pkg/controller/controller.go]: It doesn't need to be public to be unit tested. Everything is in one package.

Comment (jberkhahn) [reply pkg/controller/controller.go]: Doesn't need to be, but if it's going to be unit tested directly instead of being subsumed in the testing of another feature, it should be public.

Comment (MHBauer) [reply pkg/controller/controller.go]: I do not understand. Why? It is directly testable as a private function.

**Comment by MHBauer** [comment]
forgot the slashy thing
/approve
if you're okay with maintaining this, feel free to take off the hold
/hold
